### DISCUSSION TITLE
P4 Programming Language - lexer and parser collaboration

### DISCUSSION THREAD

**Discussion by PBug90** [tree-sitter/tree-sitter#957] [Q&A] [answered]
Quotation from https://p4.org/p4-spec/docs/P4-16-v1.2.1.html#sec-grammar:

> The grammar is actually ambiguous, so the lexer and the parser must collaborate for parsing the language. In particular, the lexer must be able to distinguish two kinds of identifiers:
> * Type names previously introduced (TYPE_IDENTIFIER tokens)
> * Regular identifiers (IDENTIFIER token)

> The parser has to use a symbol table to indicate to the lexer how to parse subsequent appearances of identifiers.

Is this even possible with tree-sitter? If not, are there any alternative approaches to parse this grammar with tree-sitter?

**Comment by maxbrunsfeld** [comment]
I converted this into a discussion thread; we're trying to move these types of Q&A issues into this forum, so that we don't ever need to "close them", and they be more discoverable to other users.

**Comment by maxbrunsfeld** [comment, marked as answer]
> the lexer and the parser must collaborate for parsing the language. In particular, the lexer must be able to distinguish two kinds of identifiers

C and C++ have the same property. We deal with it by approximating the language, using ambiguity. Our grammar contains rules something like this:

```js
rules: {
  _expression: $ => choice(
    $.identifier,
    // ...
  ),

  _type: $ => choice(
    alias($.identifier, $.type_identifier),
  ),

  // ...
}
```

In places where both a *type* and an *expression* would be valid, an ambiguity occurs: Tree-sitter doesn't know how to interpret the `identifier`. We opt-into this ambiguity at compile time using a *conflict*:

```js
conflicts: $ => [
  [$._expression, $._type],
  // ...
]
```

Often, most types will be distinguishable from expressions by looking at the token that comes afterward, so Tree-sitter won't have to *choose* between multiple valid parses. But sometimes, there are constructs that would be valid as types *and* as expression.

For example, in C, the code `a * b` can parse in two ways:

1. If `a` is a type, then like this:

    ```
    (declaration
      (type_identifier)
      (pointer_declarator
        (identifier)))
    ```
2. If `a` is not a type, then like this:

    ```
    (expression_statement
      (binary_expression
        (identifier)
        (identifier)))
    ```

In the case of C, we make the approximation that case 1 is far more common than case 2, so we can make Tree-sitter choose case 1 using **dynamic precedence**:

```js
rules: {
  pointer_declarator: $ => prec.dynamic(1, seq(
    '*',
    $._declarator
  ))
}
```
